### PR TITLE
Improve lima 2D controller + add 1D controller

### PR DESCRIPTION
**Pull request by tiagocoutinho**
This PR proposes:
* a sardana 1D controller for a lima detector
* working versions of both 1D and 2D controller with `ValueRefEnabled = False`

### CODE REVIEW

**Comment by rhomspuron** [comment]
Nice :)

One question, Does this PR solve the issue #5? If not, maybe it is a good moment to do it.

File: sardana_limaccd/ctrl/LimaCCDCtrl.py [hunk 1]
```diff
@@ -58,121 +63,144 @@
...
+
+
+class LimaImageFormat(struct.Struct):
+
+    Magic = 0x44544159
+    DTypes = ('u1', 'u2', 'u4', None, 'i1', 'i2', 'i4')
+    DTypeSize = (1, 2, 4, None, 1, 2, 4)
+
+    def __init__(self):
+        super().__init__('<IHHIIHHHHHHHHIIIIIIII')
```
Comment (jordiandreu) [inline sardana_limaccd/ctrl/LimaCCDCtrl.py]: Should this controllers be python2 compatible? In this case, we need to fix the super calls, adding the class and self instance as arguments.

**Review by jordiandreu — CHANGES_REQUESTED** [review]

File: sardana_limaccd/ctrl/LimaCCDCtrl.py (hunk 1 above)
Comment (rhomspuron) [reply sardana_limaccd/ctrl/LimaCCDCtrl.py]: Yes, the idea is to have the same version on py2 and py3 if it does not block the development.

**Comment by jordiandreu** [comment]
I'm currently also testing the code in sardana-py3 (also in python2) and there is an exception that need to be addressed:
`  File "/home/jandreu/Development/ALBA-CS/sardana/src/sardana/pool/poolbasechannel.py", line 446, in extend_value_ref_buffer
    if len(value_refs) == 0:
TypeError: object of type 'SardanaValue' has no len()`

File: sardana_limaccd/ctrl/LimaCCDCtrl.py (hunk 1 above)
Comment (tiagocoutinho) [reply sardana_limaccd/ctrl/LimaCCDCtrl.py]: Ok I can do that.
I completely forgot this was to be compatible with py2

**Comment by tiagocoutinho** [comment]
@rhomspuron :
Thanks, I plan to use it within ALBA MSPD to control the mythen detector.
Actually, this fixes #5 in the new supported feature of `ValueRefEnabled=False`.
I propose that to fix #5 we create another PR. I have some idea to fix it but it needs to be tested with old a new versions of Lima

**Comment by tiagocoutinho** [comment]
@jordiandreu ,
Thanks for testing!

> I'm currently also testing the code in sardana-py3 (also in python2) and there is an exception that need to be addressed:
File "/home/jandreu/Development/ALBA-CS/sardana/src/sardana/pool/poolbasechannel.py", line 446, in extend_value_ref_buffer if len(value_refs) == 0: TypeError: object of type 'SardanaValue' has no len()

I suspect a bug in sardana...

**Comment by tiagocoutinho** [comment]
@jordiandreu,
The error you got was for Lima2D controller or for a Lima1D controller? Were you testing with `ValueRefEnabled=False`?

**Comment by jordiandreu** [comment]
> @jordiandreu,
> The error you got was for Lima2D controller or for a Lima1D controller? Were you testing with `ValueRefEnabled=False`?

@tiagocoutinho ,
I'm using Lima2D controller with `ValueRefEnabled=True`
The measurement group has a single channel, which corresponds to the first axis of the Lima2D controller.

**Comment by reszelaz** [comment]
> I'm using Lima2D controller with ValueRefEnabled=True
The measurement group has a single channel, which corresponds to the first axis of the Lima2D controller.

In principle, this should work, at least for the dummy 2D controller it worked not so long time ago, see example from https://github.com/sardana-org/sardana/pull/1510#issuecomment-786265662.

**Comment by tiagocoutinho** [comment]
@jordiandreu ,
which type of synchronization do you have?
I think I can reproduce your error in SoftwareStart

**Comment by jordiandreu** [comment]
> @jordiandreu ,
> which type of synchronization do you have?
> I think I can reproduce your error in SoftwareStart

Indeed, I was using Software synchronization and Start trigger. Moreover, a similar problem arises by using a dummy trigger gate controller as synchronizer and Start trigger...

**Comment by rhomspuron** [comment]
> Indeed, I was using Software synchronization and Start trigger. Moreover, a similar problem arises by using a dummy trigger gate controller as synchronizer and Start trigger...

If you use a dummy trigger, it will fail because Lima will wait for HW trigger.

**Comment by tiagocoutinho** [comment]
As discussed with @rhomspuron and @jordiandreu I change this PR to include all improvements including adding a 1D controller.

## Improvements

* [x] Add MasterTrigger property (default=False): allows to set the detector as master trigger
* [x] Check detector capabilites
* [x] Use internal trigger multi when detector supports it or internal trigger otherwise
* [x] Fix latency calculation based on synchro mode
* [x] Fix moving state on startup
* [x] In prepare, stop last acquisition if it did not finish well (ex: lost of a trigger)
* [x] Fix MaxDimSize
* [x] Only allow 2D to work in `ValueRefEnabled=True`

## 1D Controller

* [x] Support ALBA use cases: mythen (dectris) and Xspress3 (quantum detectors)
* [x] Support multiple axes. Each axis number corresponds to 2D row index
* [x] Only allow 1D to work in `ValueRefEnabled=False` (at least until [sardana #1445](https://github.com/sardana-org/sardana/issues/1445) is fixed

**Comment by tiagocoutinho** [comment]
> In principle, this should work, at least for the dummy 2D controller it worked not so long time ago, see example from sardana-org/sardana#1510 (comment).

I confirm that it was a problem with my implementation.

File: sardana_limaccd/ctrl/LimaCCDCtrl.py (hunk 1 above)
Comment (tiagocoutinho) [reply sardana_limaccd/ctrl/LimaCCDCtrl.py]: Ok, latest commit should fix plenty of things including this one

**Comment by tiagocoutinho** [comment]
I have tested with a simulated mythen and the official Lima simulator in software start and software trigger. 
It seems to work.

I have pending to test:
* mythen on the beamline with hardware trigger. 
* environment with saving to NFS to check saving still works well in "slow" file systems

**Comment by tiagocoutinho** [comment]
Currently, the latest `lima-tango-server` conda package (1.9.2) has some bugs using python 3.
I recommend using directly the latest from [lima tango server repo](https://github.com/esrf-bliss/Lima-tango-python) for testing.

Namely:
* readImageSeq does not work in python 3
* if detector has capabilities like shutter, the pytango 9 server fails to start

File: sardana_limaccd/client.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+
+class LimaImageFormat(struct.Struct):
+
+    Magic = 0x44544159
+    DTypes = ("u1", "u2", "u4", None, "i1", "i2", "i4")
+    DTypeSize = (1, 2, 4, None, 1, 2, 4)
+
+    def __init__(self):
+        super().__init__("<IHHIIHHHHHHHHIIIIIIII")
```
Comment (jordiandreu) [inline sardana_limaccd/client.py]: This has to be:
`super(LimaImageFormat, self).__init__("<IHHIIHHHHHHHHIIIIIIII")`

File: sardana_limaccd/client.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+            self.config = {"saving_mode": "MANUAL"}
+            self.lima["saving_mode"] = "MANUAL"
+            return
+        self.config = config = saving_for_pattern(self.pattern)
+        config["saving_mode"] = "AUTO_FRAME"
+        config["saving_overwrite_policy"] = "ABORT"
+        names, values = zip(*config.items())
+        self.lima[names] = values
+        if self.first_image_nb != 0:
+            self.lima["saving_index_next_number"] = -1
```
Comment (jordiandreu) [inline sardana_limaccd/client.py]: The correct key is saving_next_number
`self.lima["saving_next_number"] = -1`

File: sardana_limaccd/client.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            time.sleep(0.05) # empirical value?
+            self.lima["saving_prefix"] = config["saving_prefix"]
+            t0 = time.monotonic()
+            saving_next_number = -1
+            # After settinge the prefix with saving mode = ABORT,
```
Comment (jordiandreu) [inline sardana_limaccd/client.py]: just a typo here

**Review by jordiandreu — CHANGES_REQUESTED** [review]
By now, tested in sardana develop with python3.7. More test with python 2.7 ongoing.

File: sardana_limaccd/client.py (hunk 2 above)
Comment (tiagocoutinho) [reply sardana_limaccd/client.py]: Agree. nice catch

File: sardana_limaccd/client.py (hunk 3 above)
Comment (tiagocoutinho) [reply sardana_limaccd/client.py]: Agree. Good one

File: sardana_limaccd/client.py (hunk 4 above)
Comment (tiagocoutinho) [reply sardana_limaccd/client.py]: Ok, should be fixed

File: sardana_limaccd/client.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+        self.config = config = saving_for_pattern(self.pattern)
+        config["saving_mode"] = "AUTO_FRAME"
+        config["saving_overwrite_policy"] = "ABORT"
+        names, values = zip(*config.items())
+        self.lima[names] = values
+        if self.first_image_nb != 0:
+            self.lima["saving_next_number"] = -1
+            time.sleep(0.05) # empirical value?
+            self.lima["saving_prefix"] = config["saving_prefix"]
+            t0 = time.monotonic()
```
Comment (jordiandreu) [inline sardana_limaccd/client.py]: `time.monotonic` is not available for python < 3.3. Some third party modules implement this method for old python version (https://pypi.org/project/monotonic/) but maybe is better to do not add an extra dependency if can be easily avoided.

File: sardana_limaccd/client.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            # After setting the prefix with saving mode = ABORT,
+            # the LimaCCDs takes some seconds to update the saving next
+            # number, this time depends of the number of files on the folder
+            # with the same pattern. For that reason the controller will
+            # change the first saving next number on the first start.
+
+            # Allow to set the First Image Number to any value different to
+            # 0, default value on LimaCCDs after writing the prefix with
+            # saving mode in Abort
+            while saving_next_number == -1 and time.monotonic() - t0 < 2.5:
```
Comment (jordiandreu) [inline sardana_limaccd/client.py]: Also here...

**Review by jordiandreu — CHANGES_REQUESTED** [review]

File: sardana_limaccd/client.py (hunk 5 above)
Comment (tiagocoutinho) [reply sardana_limaccd/client.py]: Yes, you are right.

I guess we can do something simple like:
```python
 monotonic = getattr(time, "monotonic", time.time)
```

Comment (jordiandreu) [reply sardana_limaccd/client.py]: Indeed, once ready I can test the code again.

File: sardana_limaccd/client.py (hunk 6 above)
Comment (jordiandreu) [reply sardana_limaccd/client.py]: Done!

**Review by jordiandreu — APPROVED** [review]

**Comment by tiagocoutinho** [comment]
I had to add a protection for the case of hardware trigger (start/stop/gate) where there are multiple starts (ex: in a step scan)

I am sorry but I have to ask for another review.

**Comment by tiagocoutinho** [comment]
Here are the tests I did today in the lab using a mythen 2 as a lima 1D controller:

ValueRefEnabled = FALSE | (which means lima saving = False; all frames stored by sardana)

source | type | Lima eq. | ct | ascan | timescan | ascanc
-- | -- | -- | -- | -- | -- | --
hw (tg1) | trigger | ext_trig_mult | OK | OK | OK | OK
hw (tg1) | start | ext_trig | OK | --- | OK | OK
hw (tg1) | gate | ext_gate | TODO (1)
soft | trigger | int_trig_mult | OK | OK | OK | OK
soft | start | int_trig | OK | --- | OK | OK
soft | gate | int_trig_mult | OK | OK | OK | OK

(1) I need to wait for a proper cable

**Review by jordiandreu — APPROVED** [review]
Nice work! I have validated this version using python2/python3 Sardana versions with a Lima Simulator 2D controller, with value_ref_enable=True and macros **ascanct** and **ct**.

**Comment by tiagocoutinho** [comment]
Thanks for yet another review @jordiandreu.
On my side I am satisfied with the results. I don't expect to make any more changes.
I guess we could consider merging it now

**Review by rhomspuron — COMMENTED** [review]
Nice work!!

Only one question:
The LimaCtrl.py has the same code than LimaCCDCtrl with different classes names, why we would like to have this?

**Comment by tiagocoutinho** [comment]
Nice catch @rhomspuron. LimaCtrl.py is just garbage that got into git by accident.
I just removed it

**Comment by tiagocoutinho** [comment]
Ok, if nobody disagrees I can merge it then.

**Review by rhomspuron — APPROVED** [review]

**Pull request merged**